core/src/main/java/org/apache/iceberg/FileScanTaskParser.java

```diff
@@ -0,0 +1,128 @@
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import com.fasterxml.jackson.core.JsonGenerator;
+import com.fasterxml.jackson.databind.JsonNode;
+import java.io.IOException;
+import org.apache.iceberg.expressions.Expression;
+import org.apache.iceberg.expressions.ExpressionParser;
+import org.apache.iceberg.expressions.Expressions;
+import org.apache.iceberg.expressions.ResidualEvaluator;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.ImmutableList;
+import org.apache.iceberg.util.JsonUtil;
+
+public class FileScanTaskParser {
+  private static final String SCHEMA = "schema";
+  private static final String SPEC = "spec";
+  private static final String DATA_FILE = "data-file";
+  private static final String DELETE_FILES = "delete-files";
+  private static final String RESIDUAL = "residual-filter";
+
+  private FileScanTaskParser() {}
+
+  public static String toJson(FileScanTask fileScanTask) {
+    return JsonUtil.generate(
+        generator -> FileScanTaskParser.toJson(fileScanTask, generator), false);
+  }
+
+  private static void toJson(FileScanTask fileScanTask, JsonGenerator generator)
+      throws IOException {
+    Preconditions.checkArgument(fileScanTask != null, "Invalid file scan task: null");
+    Preconditions.checkArgument(generator != null, "Invalid JSON generator: null");
+    generator.writeStartObject();
+
+    generator.writeFieldName(SCHEMA);
+    SchemaParser.toJson(fileScanTask.schema(), generator);
+
+    generator.writeFieldName(SPEC);
+    PartitionSpec spec = fileScanTask.spec();
+    PartitionSpecParser.toJson(spec, generator);
+
+    if (fileScanTask.file() != null) {
+      generator.writeFieldName(DATA_FILE);
+      ContentFileParser.toJson(fileScanTask.file(), spec, generator);
+    }
+
+    if (fileScanTask.deletes() != null) {
+      generator.writeArrayFieldStart(DELETE_FILES);
+      for (DeleteFile deleteFile : fileScanTask.deletes()) {
+        ContentFileParser.toJson(deleteFile, spec, generator);
+      }
+      generator.writeEndArray();
+    }
+
+    if (fileScanTask.residual() != null) {
+      generator.writeFieldName(RESIDUAL);
+      ExpressionParser.toJson(fileScanTask.residual(), generator);
+    }
+
+    generator.writeEndObject();
+  }
+
+  public static FileScanTask fromJson(String json, boolean caseSensitive) {
+    Preconditions.checkArgument(json != null, "Invalid JSON string for file scan task: null");
+    return JsonUtil.parse(json, node -> FileScanTaskParser.fromJsonNode(node, caseSensitive));
+  }
+
+  private static FileScanTask fromJsonNode(JsonNode jsonNode, boolean caseSensitive) {
+    Preconditions.checkArgument(jsonNode != null, "Invalid JSON node for file scan task: null");
+    Preconditions.checkArgument(
+        jsonNode.isObject(), "Invalid JSON node for file scan task: non-object (%s)", jsonNode);
+
+    JsonNode schemaNode = jsonNode.get(SCHEMA);
```

Comment:
probably better to change this to `JsonUtil.get(SCHEMA, jsonNode)` to have proper error reporting. Also this can then be inlined with the next line